### DISCUSSION TITLE
Improve type inference

### DISCUSSION THREAD

**Discussion by beikov** [hibernate/hibernate-orm#4747] [Design Proposals]
Right now, we always try to determine the `MappingModelExpressible` in `BaseSqmToSqlAstConverter` by looking at an SQM node, but for certain expressions/predicates e.g. binary arithmetic or comparison predicate, we could instead make use of `Expression#getExpressionType` of the "other side". For example, if we first translate the LHS and then the RHS of a comparison predicate like `path = :param`, it would be much better to use the expression type of the LHS to infer the type of the RHS.
A possible downside of this would be, that we probably wouldn't figure out issues in the `determineValueMapping` function so fast because usually the LHS in expressions/predicates is a typed node.

Any reason for not applying this improvement?

We currently do something like this:

```java
inferrableTypeAccessStack.push( () -> determineValueMapping( rightOperand ) );
final Expression lhs = toSqlExpression( leftOperand.accept( this ) );
inferrableTypeAccessStack.pop();

inferrableTypeAccessStack.push( () -> determineValueMapping( leftOperand ) );
final Expression rhs = toSqlExpression( rightOperand.accept( this ) );
inferrableTypeAccessStack.pop();
```

whereas we could do this:

```java
inferrableTypeAccessStack.push( () -> determineValueMapping( rightOperand ) );
final Expression lhs = toSqlExpression( leftOperand.accept( this ) );
inferrableTypeAccessStack.pop();

inferrableTypeAccessStack.push( () -> lhs.getExpressionType() );
final Expression rhs = toSqlExpression( rightOperand.accept( this ) );
inferrableTypeAccessStack.pop();
```

**Comment by sebersole** [comment]
Really not sure what you are proposing here exactly, so hard to say.  

Should we "improve" things where there is an opportunity?  Yes, of course.  Kind of depends on details.

**Reply by beikov** [reply]
I added an example to the description above now.

**Reply by beikov** [reply]
It doesn't seem worth it. I'll not do that for now.
